### PR TITLE
Support forward https protocol port

### PR DESCRIPTION
**Pull request by iQQBot**
## Description

Support "Change port protocol" (HTTP/HTTPS)

Some user using `https` for their service, but now Gitpod only support forward `http` protocol, which means these service cannot visit in workspace ports url, they only can test it via local-desktop IDE (ssh port-forward)

This PR also changes supervisor port forward using nftable, it support any protocol, more fast and easy.

## Related Issue(s)

Part of #15915

## How to test

1. start a workspace with this repo https://github.com/iQQBot/test.test/tree/https in [preview env](https://pd-port-15915.preview.gitpod-dev.com/workspaces)
2. it should able access port 5173 via workspace port url
3. use `curl https://localhost:5173 -v -k` make sure protocol is https
4. you can also using `gp ports protocol 5173:https` to change protocol
5. or define port protocol in `.gitpod.yml`
6. change `vite.config.ts`, remove line `host: '0.0.0.0',` and restart vite, make sure workspace port URL still can work.

if you run `sudo lsof -i :5173` you should see two listen, the one is from node, the other one is from supervisor
image

if you close node server, run `sudo lsof -i :5173` again, you should see clean list
image

## Release Notes

```release-note
Support forward https protocol port
```

## Build Options:

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`

<details>
<summary>Publish Options</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer Options</summary>

- [ ] with-ee-license
- [ ] with-dedicated-emulation
- [x] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

#### Preview Environment Options:
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

**Comment by jeanp413** [comment]
Can also be tested with:
- creating a key and cert with `openssl req -newkey rsa:2048 -new -nodes -x509 -days 3650 -keyout key.pem -out cert.pem`
- run `npx http-server -S`
- play with `gp ports protocol` command

**Comment by jeanp413** [comment]
Works as expected :tada: :rocket:

**Comment by kylos101** [comment]
/hold

@laushinka can we talk about this from a prioritization standpoint?

**Comment by laushinka** [comment]
> /hold
> 
> @laushinka can we talk about this from a prioritization standpoint?

@kylos101 Let's. I'll add it to our Monday sync?

**Comment by kylos101** [comment]
> @kylos101 Let's. I'll add it to our Monday sync?

Perfect, thank you @laushinka ! :pray:

**Comment by kylos101** [comment]
:wave: @iQQBot , @laushinka and I spoke about this PR, Workspace Team is focused on MK2 this week (we're at a key point). Reviews for this PR will likely be delayed till next week. Hope that's okay?

**Review by akosyakov — APPROVED** [review]

**Comment by akosyakov** [comment]
@iQQBot Should it be supported from the ports view as well? Should we update docs? cc @loujaybee

**Comment by kylos101** [comment]
> Hey @kylos101 does the workspace team have time to review this PR this week?

Dear @iQQBot , I'm sorry, at the moment, no. We're having to be selective at the moment (we have many PRs and issues in-flight for Dedicated). If I were to guess, I think next week would be better/possible.

**Comment by easyCZ** [comment]
Moving to draft as this has gone extremely stale. If it's still needed, please rebase and put for review.

**Comment by akosyakov** [comment]
What is blocking a review here for 3 months? Changes seem to be trivial.

File: components/ws-proxy/pkg/proxy/routes.go [hunk 1]
```diff
@@ -434,6 +437,11 @@ func installWorkspacePortRoutes(r *mux.Router, config *RouteHandlerConfig, infoP
 				workspacePodPortResolver,
 				withHTTPErrorHandler(showPortNotFoundPage),
 				withXFrameOptionsFilter(),
+				func(h *proxyPassConfig) {
+					h.Transport = &http.Transport{
```
Comment (akosyakov) [inline components/ws-proxy/pkg/proxy/routes.go:441]: @iQQBot is it possible to make it implicit? i.e. try to do http request first and then it fails with TLS then try https? cc @csweichel 

@loujaybee from product perspective would it be better to have explicit configuration similar to visibility as Codespaces do, or implicit detection would be good enough?

Comment (akosyakov) [reply components/ws-proxy/pkg/proxy/routes.go:441]: Are there other ways to detect?

Comment (akosyakov) [reply components/ws-proxy/pkg/proxy/routes.go:441]: ok, asking if we can reduce the scope of the PR, but if not then it is alright too 👍

**Comment by kylos101** [comment]
@iQQBot a bumped into a couple issues:

1. HTTPS exposed traffic works fine. However, I ran `curl lama.sh | sh` in a workspace from this branch's preview, but when I browsed to it, it did not work. To test with another app, I set `https: false,` in `vite.config.ts`, and got the same behavior. Suggestion: :bulb: Add an integration test, to assert that both https and http exposed ports are both working.
2. When I mouse over protocol in `.gitpod.yml`, I get `The 'protocol' property is deprecated.` Would you expect that to no longer be deprecated? I was, because you made schema changes.

**Comment by kylos101** [comment]
Thanks, @iQQBot ! Sounds like my test setup was poor, I'll try again, thank you for the clarification. :pray: 

> are you run curl lama.sh | sh in https://github.com/iQQBot/test.test/tree/https this test repo?

Yes.

>  if yes, this is expect you will see a blank page, it because at that folder it's has index.html, lama will read this file first if index.html exist it don't show folder list.

:facepalm: I was expecting a folder list, and will assert that I'm getting a HTTP 200 with the index.html.

> If you set https: false, in vite.config.ts and don't change protocol via gp cli it's expect to see something like Port 5173 Not Found it because in .gitpod.yml 5173 is set to https, so it will try connect port via https.

Oh! I see. This is new for me, changing our port config. Is through the `.gitpod.yml` the only way for now?

**Comment by kylos101** [comment]
LGTM. Lemme know once it comes out of draft and is ready for review.

**Comment by loujaybee** [comment]
> Should we update docs? - @akosyakov 

docs + changelog: https://github.com/gitpod-io/website/pull/3760

**Pull request merged**